New Multinode Kubernetes Test

pbx0 opened the pull request:
This replaces the old multinode test we had running based off some very old and unoffcial configs. The cluster is now setup according to the coreos manual setup guides in part by using the scripts from https://github.com/coreos/coreos-kubernetes/tree/master/multi-node/generic. 

Right now we also have the kubelet running via rkt fly. Once that script is finalized and accepted in https://github.com/coreos/coreos-overlay/pull/1750 I'll update this PR to use the on-disk script.

mischief commented on platform/platform.go:
```diff
@@ -128,6 +128,74 @@ func (t *TestCluster) DropFile(localPath string) error {
 }
 
+// Copy a file between two machines in a cluster.
+func TransferFile(src Machine, srcPath string, dst Machine, dstPath string) error {
```
can this be turned into e.g. `func ReadFile(m Machine, path string) (io.ReadCLoser, error)` and combined by the caller with e.g.:

```go
rc, _ := ReadFile(fromM, "/foo")
defer rc.Close()
err = InstallFile(rc, toM, "/bar")
```

pbx0 replied:
This makes sense, but, how do we avoid leaking sessions with this interface?

mischief replied:
i suppose you need a io.ReadCloser wrapper around `ssh cat` that cleans up ssh on `Close()`

mischief replied:
if you do make this change can you make it in a seperate commit?

mischief commented:
i think instances of `strings.Replace` should be using go `text/template` pkg instead.

aaronlevy commented on kola/tests/kubernetes/controllerInstall.go:
```diff
@@ -0,0 +1,795 @@
+export ETCD_ENDPOINTS=
+
+# Specify the version (vX.Y.Z) of Kubernetes assets to deploy
+export K8S_VER=v1.1.3
```
This ideally is parameterized in some way (also we very likely start by publishing images of v1.1.2, v1.1.4, and v1.1.7)

mischief replied:
if go `text/template` is used, we can pass it thru the template and maybe test multiple versions of k8s.

pbx0 replied:
I've parameterized this (locally) to single version variable in main. I'm not sure the best way to pass that into our tests since this will be run on os releases as well. 

Maybe we just maintain a list of the versions we want to test and run a cluster for each version?

mischief replied:
similar style to how https://github.com/coreos/mantle/blob/master/kola/tests/flannel/flannel.go templates different flannel backends? can have tests like `"google.kubernetes.coreosbasic.v1.1.2"` `"google.kubernetes.coreosbasic.v1.2"`

aaronlevy commented on kola/tests/kubernetes/coreosbasic.go:
```diff
@@ -0,0 +1,319 @@
+	return nil
+}
+
+func generateWorkerTLSAssets(master platform.Machine, workers []platform.Machine) error {
```
Is there a reason it's necessary to copy around CA assets then ssh/run openssl? This could be done directly in go with crypto/x509

pbx0 replied:
We could probably use crypto/x509 but it'd be far more verbose and we'd have to copy all the files to the machines anyway.

Since this test follows our install guides somewhat closely, I think its less of a burden to maintain if we just shell out to the same openssl commands.

aaronlevy commented on kola/tests/kubernetes/coreosbasic.go:
```diff
@@ -0,0 +1,319 @@
+
+	// configure kubectl
+	if err := configureKubectl(master, MASTER_HOST); err != nil {
+		return nil
```
return err?

pbx0 replied:
done

aaronlevy commented on kola/tests/kubernetes/coreosbasic.go:
```diff
@@ -0,0 +1,319 @@
+
+	// check that all nodes appear in kubectl
+	f := func() error {
+		if err = nodeCheck(master, workers); err != nil {
```
nit: redundant if err checks.
```
f := func() {
    return nodeCheck(master, workers)
}
```

pbx0 replied:
done

aaronlevy commented on kola/tests/kubernetes/coreosbasic.go:
```diff
@@ -0,0 +1,319 @@
+		CA_CERT    = "/home/core/ca.pem"
+		ADMIN_KEY  = "/home/core/admin-key.pem"
+		ADMIN_CERT = "/home/core/admin.pem"
+		kubeURL    = "https://storage.googleapis.com/kubernetes-release/release/v1.1.3/bin/linux/amd64/kubectl"
```
Should tie the version of kubectl to the rest of assets / test

pbx0 replied:
done

pbx0 commented:
@aaronlevy @mischief PTAL

Fixed some small things, used text/template, and there is now a central list of version tags that are all registered. 

I also added a commit that refactors the transfer function. I am lukewarm about the results since I don't think ReadFile is going to be very useful by itself.

mischief commented:
is this test going to require the kubelet script that @aaronlevy and @pbx0 just wrote?

if so, the test should specify the min version of coreos required is the one that has that script.

pbx0 commented:
It includes it for the time being, but the idea is that it will depend on the on-disk version, of course.

We could merge that PR and specify min version or we could just skip the test if we can't stat the on-disk script.

mischief commented:
ok, the merged code should certainly include some kind of check.

probably should allow the Skip code to include a message indicating why the test was skipped.

pbx0 commented:
@mischief PTAL.

Since this commit deletes the previous kubernetes test, I decided to not skip when it can't find the on-disk kubelet script. Instead, if the on-disk doesn't exist, it will fallback to using the test's built-in version.  The scripts are currently identical and we can eventually remove this check.

mischief commented:
if @aaronlevy has no more complaints lgtm.

aaronlevy commented:
lgtm
